test/components/generators/chat/test_hugging_face_local.py

```diff
@@ -307,3 +353,124 @@ def test_live_run(self, monkeypatch):
+        assert message.meta["finish_reason"] == "tool_calls"
+
+    def test_run_with_tools_in_run_method(self, model_info_mock, tools):
+        generator = HuggingFaceLocalChatGenerator(model="meta-llama/Llama-2-13b-chat-hf")
+
+        # Mock pipeline and tokenizer
+        mock_pipeline = Mock(return_value=[{"generated_text": '{"name": "weather", "arguments": {"city": "Paris"}}'}])
+        mock_tokenizer = Mock(spec=PreTrainedTokenizer)
+        mock_tokenizer.encode.return_value = ["some", "tokens"]
+        mock_tokenizer.pad_token_id = 100
+        mock_tokenizer.apply_chat_template.return_value = "test prompt"
+        mock_pipeline.tokenizer = mock_tokenizer
+        generator.pipeline = mock_pipeline
+
+        messages = [ChatMessage.from_user("What's the weather in Paris?")]
+        results = generator.run(messages=messages, tools=tools)
+
+        assert len(results["replies"]) == 1
+        message = results["replies"][0]
+        assert message.tool_calls
+        tool_call = message.tool_calls[0]
+        assert isinstance(tool_call, ToolCall)
+        assert tool_call.tool_name == "weather"
+        assert tool_call.arguments == {"city": "Paris"}
+        assert message.meta["finish_reason"] == "tool_calls"
+
+    def test_run_with_tools_and_tool_response(self, model_info_mock, tools):
+        generator = HuggingFaceLocalChatGenerator(model="meta-llama/Llama-2-13b-chat-hf")
+
+        # Mock pipeline and tokenizer
+        mock_pipeline = Mock(return_value=[{"generated_text": "The weather in Paris is 22°C"}])
+        mock_tokenizer = Mock(spec=PreTrainedTokenizer)
+        mock_tokenizer.encode.return_value = ["some", "tokens"]
+        mock_tokenizer.pad_token_id = 100
+        mock_tokenizer.apply_chat_template.return_value = "test prompt"
+        mock_pipeline.tokenizer = mock_tokenizer
+        generator.pipeline = mock_pipeline
+
+        tool_call = ToolCall(tool_name="weather", arguments={"city": "Paris"})
+        messages = [
+            ChatMessage.from_user("What's the weather in Paris?"),
+            ChatMessage.from_assistant(tool_calls=[tool_call]),
+            ChatMessage.from_tool(tool_result="22°C", origin=tool_call),
+        ]
+        results = generator.run(messages=messages)
+
+        assert len(results["replies"]) == 1
+        message = results["replies"][0]
+        assert not message.tool_calls  # No tool calls in the final response
+        assert "22°C" in message.text
+        assert message.meta["finish_reason"] == "stop"
+
+    def test_run_with_custom_tool_parser(self, model_info_mock, tools):
+        """Test that a custom tool parsing function works correctly."""
+        generator = HuggingFaceLocalChatGenerator(
+            model="meta-llama/Llama-2-13b-chat-hf", tools=tools, tool_parsing_function=custom_tool_parser
+        )
+        generator.pipeline = Mock(return_value=[{"generated_text": "Let me check the weather for you"}])
```

Comment:
this `generated_text` is confusing. Please use something like `"mock response. We don't use it"`